tests/functionaltests/Tests/NSURLConnection.mm

```diff
@@ -240,37 +300,129 @@ static void _testRequestWithURL(NSURLConnectionTestHelper* connectionTestHelper)
 
     // Make sure we received a response.
     ASSERT_EQ_MSG(NSURLConnectionDelegateDidReceiveResponse,
                   [(NSNumber*)[connectionTestHelper.delegateCallOrder objectAtIndex:0] integerValue],
-                  "FAILED: didReceiveResponse should be the first delegate to be called!");
+                  "FAILED: didReceiveResponse should be the 1st delegate to be called!");
     ASSERT_TRUE_MSG((connectionTestHelper.response != nil), "FAILED: Response cannot be empty!");
+
     NSURLResponse* response = connectionTestHelper.response;
-    if (![response isKindOfClass:[NSHTTPURLResponse class]]) {
-        ASSERT_FALSE_MSG(true, "FAILED: Response should be of kind NSHTTPURLResponse class!");
-    }
+    ASSERT_TRUE_MSG([response isKindOfClass:[NSHTTPURLResponse class]], "FAILED: Response should be of kind NSHTTPURLResponse class!");
+
     NSHTTPURLResponse* httpResponse = (NSHTTPURLResponse*)response;
     LOG_INFO("Received HTTP response status: %ld", [httpResponse statusCode]);
     ASSERT_EQ_MSG(404, [httpResponse statusCode], "FAILED: HTTP status 404 was expected!");
 
+    // Make sure we received a didFinishLoading callback
+    ASSERT_EQ_MSG(NSURLConnectionDelegateDidFinishLoading,
+                  [(NSNumber*)[connectionTestHelper.delegateCallOrder objectAtIndex:1] integerValue],
+                  "FAILED: didReceiveData should be the 2nd delegate to be called!");
+
     // Make sure we did not receive any data.
     ASSERT_TRUE_MSG((connectionTestHelper.data == nil), "FAILED: We should not have received any data!");
 
     // Make sure there was no error.
     ASSERT_TRUE_MSG((connectionTestHelper.error == nil), "FAILED: Connection returned error %@!", connectionTestHelper.error);
 }
 
+static void _testRequestWithURL_Post(NSURLConnectionTestHelper* connectionTestHelper, bool useStream) {
+    NSString* urlString = @"https://httpbin.org/post";
+    LOG_INFO("Establishing upload task with url %@", urlString);
+
+    NSMutableURLRequest* request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:urlString]];
+
+    // Generate a random alphanumeric string to use in the body
+    static const std::string alphanumeric = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+
+    std::vector<unsigned char> data((1 << 15) + 1);
+    std::default_random_engine engine;
+    engine.seed(2);
+    std::uniform_int_distribution<> distribution(0, alphanumeric.length() - 1);
+    std::generate(data.begin(), data.end(), [&distribution, &engine] { return alphanumeric[distribution(engine)]; });
+
+    NSData* nsData = [NSData dataWithBytesNoCopy:data.data() length:data.size() freeWhenDone:NO];
+
+    if (useStream) {
+        request.HTTPBodyStream = [NSInputStream inputStreamWithData:nsData];
+    } else {
+        request.HTTPBody = nsData;
+    }
+    request.HTTPMethod = @"POST";
+    [request setValue:[NSString stringWithFormat:@"%lu", data.size()] forHTTPHeaderField:@"Content-Length"];
```

Comment:
The burden might be on `NSURLProtocol` to set the outgoing content-length, right?